Async Workers

mxgrey opened the pull request:
This PR is the final culmination of all the async execution features that I currently have plans to implement. This builds on top of #421 so there is no need to review this PR until that one is merged, but I'm opening this PR to help make it visible to anyone who wants to try out the full breadth of async features that I've been working on.

The key role of this PR is to introduce the worker concept that was initially proposed in [these slides](https://docs.google.com/presentation/d/1JlB2mx5hh_LxcJjPc-ognPRfxXA-5FQcrZSIi_dBPYI/edit#slide=id.g263796b55e4_0_278). The API of this implementation has some differences from the initial proposal, but the concept is exactly the same.

Since this is just a draft preview PR, in lieu of a detailed explanation, I'll direct folks to [this slide deck](https://docs.google.com/presentation/d/1TFq2i6zA4lB8VOlvgQTBs1-yuCC1Grz1Vcfxs07KxJc/edit?usp=sharing) which explains the motivation behind the Worker concept.

maspe36 commented on rclrs/src/executor.rs:
```diff
@@ -1,84 +1,466 @@
+/// This trait defines the interface for having an executor run.
+pub trait ExecutorRuntime: Send {
+    /// Get a channel that can add new items for the executor to run.
+    fn channel(&self) -> Arc<dyn ExecutorChannel>;
```
Saw this while reviewing #428, couldn't we use associated types to get away from trait objects here? I don't think we really need the extra flexibility. Do we really plan on doing something like storing executors in a heterogeneous container? (i.e. `Vec<Box<dyn ExecutorRuntime>>`)
```rust
pub trait ExecutorRuntime: Send {
    type ExecutorChannel: ExecutorChannel + Send + Sync;

    fn channel(&self) -> Arc<Self::ExecutorChannel>;

    fn spin(&mut self, conditions: SpinConditions) -> Vec<RclrsError>;

    fn spin_async(
        self,
        conditions: SpinConditions,
    ) -> BoxFuture<'static, (Self, Vec<RclrsError>)>
    where
        Self: Sized;
}
```
As long as Self is Sized, I think we can also remove the extra `Box` in the future return

mxgrey replied:
I did consider exactly what you're describing, but the `dyn ExecutorChannel` gets used inside [`ExecutorCommands`](https://github.com/mxgrey/ros2_rust/blob/6d3f7e4660cf6c82e2416139829331c9b1566b0c/rclrs/src/executor.rs#L124) which is held by many different objects including [`Node`](https://github.com/mxgrey/ros2_rust/blob/6d3f7e4660cf6c82e2416139829331c9b1566b0c/rclrs/src/node.rs#L98), [`SubscriptionExecutable`](https://github.com/mxgrey/ros2_rust/blob/6d3f7e4660cf6c82e2416139829331c9b1566b0c/rclrs/src/subscription.rs#L215), [`ServiceExecutable`](https://github.com/mxgrey/ros2_rust/blob/6d3f7e4660cf6c82e2416139829331c9b1566b0c/rclrs/src/service.rs#L208), and in the future `Action` will probably need it.

If we used an associated type, then we would need to add a generic argument for the type of the `ExecutorRuntime` to `ExecutorCommands`, which would then require us to add a generic argument for the `ExecutorRuntime` to `Node`, `SubscriptionExecutable`, `ServiceExecutable`, and `Client`. That would undermine the separation of concerns that we currently enjoy between application code that creates nodes and primitives vs application code that chooses the runtime.

With what you're proposing, to be agnostic to the choice of runtime, user code would need to be like

```rust
struct MyNodeData<E: ExecutorRuntime> {
    node: Node<E>,
    subscription: Subscription<Msg>,
    client: Client<E, Srv>,
}

fn setup_my_application<E: ExecutorRuntime>(executor: Executor<E>) -> MyNode<E> {
    let node = executor.create_node("my_node");
    let subscription = node.create_subscription("my_topic", do_something);
    let client = node.create_client("my_service");
    MyNode { node, subscription, client }
}
```

Probably the worst part of this IMO is that we'll see inconsistencies between what kind of generic parameters are needed by Node vs Subscription vs Client vs Service. And if we change some implementation detail of Subscription or Service, it's possible that we'll need to add a generic parameter for the runtime to them.

On the other hand, using a trait object gives us a clean separation between these concerns, allows all the primitives to have consistent generic parameters, and reduces the coupling between the user-facing API and the implementation details.

romainreignier commented:
Hi, these changes looks amazing!
I am struggling to use the current API in a real world node in which I need to wait for several topics to trigger other actions like async client calls or timer timeout. It is very different than how I used to do it in C++.
From the slides, it seems that the Workers would help me organize better the Rust code.
What is the current roadmap to get Workers merged?

mxgrey commented:
> I am struggling to use the current API

Just to make sure I'm understanding correctly, when you say "the current API" you mean the API that's currently on the `main` branch, right? 

If there are any potential changes to the API of this PR that can further help with your use case, I'm happy to iterate on it.

As for roadmap / timeline, there will be a [working group meeting in about two hours](https://discourse.ros.org/t/ros-2-rust-meeting-march-2025/42421), and we'll probably be talking about next steps there.

I think in general if anyone (including interested users, rather than maintainers) is able to leave reviews on https://github.com/ros2-rust/ros2_rust/pull/429 and https://github.com/ros2-rust/ros2_rust/pull/430 that may expedite the progress towards getting them merged.

mxgrey commented:
@esteve This PR should be properly ready for review now.

To clean up the version history, I've squashed the original set of commits and rebased them onto `main`.

I've also added a considerable amount of documentation throughout the modules, including usage examples and doctests. The root module now has a quick start guide with simple examples of basic usage, parameters, and workers (mostly lifted from the `examples` directory).

Timers have been moved into #480 . @romainreignier If you were relying on this branch for timers, you can retarget to the branch of #480.

mxgrey commented on rclrs/src/client.rs:
```diff
@@ -87,24 +46,263 @@ pub struct ClientState<T>
+    ///
+    /// Consider using [`Self::notify_on_service_ready`] if you want to wait
+    /// until a service for this client is ready.
+    pub fn service_is_ready(&self) -> Result<bool, RclrsError> {
```
I agree, and more generally I've been disappointed by how many places the rclrs API tosses `Results` at users. I think it will give users the impression that our API is more brittle than it really is.

What I'd like to propose is that for any function with the following criteria:
* There is no practical risk of an error happening that would be caused by user input
* There is a reasonable fallback behavior we can follow in the unlikely event of an error 
  * "Reasonable" may include panicking for situations that indicate an unrecoverable rmw or rcl error, e.g. failure to create a subscription/publisher

we provide two API functions, one that starts with `try_` and one without. The one with `try_` will return a `Result` as our current API does, and the one without will use the fallback behavior in the event of an error.

In this case it would look like

```rust
pub fn service_is_ready(&self) -> bool {
    match self.try_service_is_ready() {
        Ok(ready) => return ready,
        Err(err) => {
            log_error!(
                "rclrs.client.service_is_ready",
                "An error occurred: {err}
            );
            return false;
        }
    };
}
pub fn try_service_is_ready(&self) -> Result<bool, RclrsError> { ... }
```

We can imagine doing likewise for `create_subscription` / `try_create_subscription` and the other primitive creation methods.

For this PR I'll just introduce this concept for `service_is_ready`, but I'll start putting together a proposal on this for the next working group meeting, and we can consider introducing this change in a follow-up PR.

mxgrey commented on rclrs/src/node/node_graph_task.rs:
```diff
@@ -0,0 +1,39 @@
+// as the Node that started it is alive.
+pub(super) async fn node_graph_task(
+    mut receiver: UnboundedReceiver<NodeGraphAction>,
+    #[allow(unused)] guard_condition: GuardCondition,
```
This is mostly a matter of opinion, but I think it's better to explicitly mark a field as being something is meant to go unused to communicate very clearly that this is not a mistake or caused by a lazy refactor.

In some cases the reason it goes unused may be obvious, like if a method in a trait definition has some extra arguments you don't need, then it's obvious to a reviewer that you can't just remove the arguments as that would cause a compilation error. In those cases I don't mind using the leading underscore trick.

But this case that you're commenting on is more subtle. The code would compile if you removed the `guard_condition` argument, but it would stop working correctly because the lifecycle of the `guard_condition` argument has some side-effects. That's why I want to make sure readers receive an extremely clear signal that the argument is there on purpose despite it going unused.

mxgrey commented on rclrs/src/executor/basic_executor.rs:
```diff
@@ -0,0 +1,452 @@
+/// If you need high-throughput multi-threaded execution, then consider using
+/// a different executor.
+//
+// TODO(@mxgrey): Implement a multi-threaded executor using tokio in a downstream
```
The comment that I left here is outdated. These days I am more inclined to do what you're suggesting: Put a tokio-based executor behind a feature flag. I think either would work, but the feature flags make some aspects of maintenance easier.

The most important thing is making sure that thirdparty executors can be made outside of the rclrs crate, which I believe they currently can.

maspe36 replied:
Sounds good. I'm also interested in writing a real-time executor at some point, so glad to see the foresight here :)

mxgrey commented on rclrs/src/publisher.rs:
```diff
@@ -95,10 +95,10 @@ where
     /// Node and namespace changes are always applied _before_ topic remapping.
-    pub(crate) fn new<'a>(
-        node_handle: Arc<NodeHandle>,
+    pub(crate) fn create<'a>(
```
This is a very opinionated thing on my part, but I think many Rust users will be accustomed to both:
* `fn new` being the public-facing method for creating an instance of an object
* `fn new` usually returns a `Self`, not usually a `Arc<Self>` much less a `Result<Arc<Self>, _>`

I don't have any strong objection to using the name `new` here, but the above two points were my rationale for changing it.

I wouldn't mind adding a `try_` prefix. Coincidentally that would go well with [this comment](https://github.com/ros2-rust/ros2_rust/pull/446#discussion_r2105791629) that I wrote before reading your suggestion here.

maspe36 replied:
I'm happy to leave the function name as create, actually. I like your rationale — it's similar to what we do at work with our "rusty C++" :D

A quick search shows that Tokio does this in a few places too:
https://docs.rs/tokio/latest/tokio/?search=create

Though they're not entirely consistent:
https://docs.rs/tokio/latest/tokio/io/unix/struct.AsyncFd.html#method.try_new

maspe36 commented on rclrs/src/client.rs:
```diff
@@ -87,24 +46,263 @@ pub struct ClientState<T>
+    board: Arc<Mutex<ClientRequestBoard<T>>>,
     #[allow(unused)]
-    node: Node,
+    lifecycle: WaitableLifecycle,
```
We have comments in all of the other call-sites describing it
https://github.com/ros2-rust/ros2_rust/pull/446/files/8f898601365c7ac8166393a9a25d35130144e240#diff-1d9414c53c73d7dac3b591eea05198c58c6099cdc60255148c77332c1dc8e6f8R29

Could be worth the maintenance cost considering how subtle it really is.

maspe36 commented:
If there are no additional comments from the other maintainers this time next week I will merge this PR

mxgrey commented:
I'm not sure how the current failure in the Windows CI would be caused by this PR :thinking: 

![Screenshot from 2025-06-09 14-36-23](https://github.com/user-attachments/assets/8f678384-7d0a-4f51-a8bd-6736b1948412)
